File: src/main/java/org/distril/beengine/entity/EntityLiving.java
package org.distril.beengine.entity;

import com.nukkitx.protocol.bedrock.data.entity.EntityEventType;
import com.nukkitx.protocol.bedrock.packet.EntityEventPacket;
import org.distril.beengine.server.Server;

public abstract class EntityLiving extends Entity {

	public EntityLiving(EntityType type) {
		super(type);
	}

	@Override
	public void setHealth(float health) {
		var wasAlive = this.isAlive();
		super.setHealth(health);
		if (this.isAlive() && !wasAlive) {
			var packet = new EntityEventPacket();
			packet.setRuntimeEntityId(this.getId());
			packet.setType(EntityEventType.RESPAWN);

			Server.getInstance().broadcastPacket(this.getViewers(), packet);
		}
	}

	@Override
	public void kill() {
		if (this.isAlive()) {
			super.kill();

			// TODO: drop all items
		}
	}
}

File: src/main/kotlin/org/distril/beengine/entity/EntityLiving.kt
package org.distril.beengine.entity

import com.nukkitx.protocol.bedrock.data.entity.EntityEventType
import com.nukkitx.protocol.bedrock.packet.EntityEventPacket
import org.distril.beengine.server.Server

abstract class EntityLiving(type: EntityType) : Entity(type) {

	override var health: Float
		get() = super.health
		set(value) {
			super.health = value

			if (!this.isAlive) {
				this.kill()

				val packet = EntityEventPacket()
				packet.runtimeEntityId = this.id
				packet.type = EntityEventType.RESPAWN

				Server.broadcastPacket(this.viewers, packet)
			}
		}

	fun kill() {
		if (!this.isAlive) return

		// TODO: drop all items
	}
}
